File: examples2d/soft_bodies2.rs
//! Soft-body showcase (2D): pressurized blobs, a rope holding a weight, a jelly square and a
//! hanging chain of soft polygons.

use rapier_testbed2d::TestbedViewer;
use rapier2d::prelude::*;

pub async fn run(viewer: &mut TestbedViewer) -> anyhow::Result<()> {
    let mut world = PhysicsWorld::new();

    /*
     * Ground and walls.
     */
    world.insert(
        RigidBodyBuilder::fixed().translation(Vector::new(0.0, -0.5)),
        ColliderBuilder::cuboid(15.0, 0.5),
    );
    world.insert(
        RigidBodyBuilder::fixed().translation(Vector::new(-15.0, 5.0)),
        ColliderBuilder::cuboid(0.5, 5.0),
    );
    world.insert(
        RigidBodyBuilder::fixed().translation(Vector::new(15.0, 5.0)),
        ColliderBuilder::cuboid(0.5, 5.0),
    );

    /*
     * Pressurized blobs of various sizes.
     */
    for i in 0..5 {
        let radius = 0.6 + 0.15 * i as Real;
        let blob = SoftBodyBuilder::disk(
            Vector::new(-10.0 + i as Real * 2.5, 2.0 + i as Real),
            radius,
            24,
        )
        .softness(SpringCoefficients::new(20.0, 1.0))
        .volume_factor(1.1)
        .self_contacts(true)
        .particle_mass(0.05);
        world.insert_soft_body(blob);
    }

    /*
     * Jelly squares (corotational and Neo-Hookean elastic cells) and one with per-cell area
     * constraints.
     */
    let jelly = SoftBodyBuilder::grid(Vector::new(2.0, 1.2), Vector::splat(1.0), 6, 6)
        .cell_model(SoftBodyCellModel::Corotational)
        .material(SoftBodyMaterial {
            young_modulus: 3.0e3,
            poisson_ratio: 0.35,
            elastic_damping_ratio: 0.5,
            ..Default::default()
        })
        .particle_mass(0.2);
    world.insert_soft_body(jelly);
    let jelly = SoftBodyBuilder::grid(Vector::new(8.0, 1.2), Vector::splat(1.0), 6, 6)
        .cell_model(SoftBodyCellModel::NeoHookean)
        .material(SoftBodyMaterial {
            young_modulus: 3.0e3,
            poisson_ratio: 0.35,
            elastic_damping_ratio: 0.5,
            ..Default::default()
        })
        .particle_mass(0.2);
    world.insert_soft_body(jelly);
    let jelly = SoftBodyBuilder::grid(Vector::new(5.0, 1.2), Vector::splat(1.0), 6, 6)
        .cell_model(SoftBodyCellModel::Volume)
        .softness(SpringCoefficients::new(20.0, 1.0))
        .particle_mass(0.2);
    world.insert_soft_body(jelly);

    /*
     * A rope hanging from a fixed anchor, holding a rigid box.
     */
    let rope = SoftBodyBuilder::rope(Vector::new(8.0, 9.0), Vector::new(12.0, 9.0), 25)
        .pinned_particles([0])
        .softness(SpringCoefficients::new(40.0, 1.0))
        .particle_mass(0.05);
    let rope_handle = world.insert_soft_body(rope);
    let last_pos = world.soft_bodies[rope_handle].particle_position(24);
    let (weight, _) = world.insert(
        RigidBodyBuilder::dynamic().translation(last_pos + Vector::new(0.0, -0.4)),
        ColliderBuilder::cuboid(0.3, 0.3).density(2.0),
    );
    world.soft_bodies[rope_handle].attach_particle(24, weight, &world.bodies);

    /*
     * A stack of rigid boxes for the blobs to knock down.
     */
    for i in 0..6 {
        world.insert(
            RigidBodyBuilder::dynamic().translation(Vector::new(-4.0, 0.3 + 0.6 * i as Real)),
            ColliderBuilder::cuboid(0.3, 0.3),
        );
    }

    viewer.set_world(&mut world);
    viewer.look_at(Vec2::new(0.0, 4.0), 30.0);

    while viewer.render_frame(&mut world).await {
        if viewer.simulating() {
            world.step();
        }
    }
    Ok(())
}

File: bindings/c/testbed/examples2d/soft_bodies2.c
/* Port of examples2d/soft_bodies2.rs. */
#include "testbed.h"
#include "rapier_helpers.h"
#include "rapier_math.h"

void tbSoftBodies2(Testbed *testbed) {
    /* World. */
    R2World *world = r2NewWorld();

    /* Ground and walls. */
    {
        R2RigidBodyDesc rigidBody = r2FixedRigidBodyDesc();
        rigidBody.position.translation = r2Vector(0, -0.5);
        rigidBody.canSleep = !testbed->noSleep;
        R2ColliderDesc collider = r2CuboidColliderDesc(r2Vector(15, 0.5));

        R2RigidBodyHandle rigidBodyHandle = r2InsertRigidBody(world, &rigidBody);
        r2InsertCollider(rigidBodyHandle, &collider);
    }
    {
        R2RigidBodyDesc rigidBody = r2FixedRigidBodyDesc();
        rigidBody.position.translation = r2Vector(-15, 5);
        rigidBody.canSleep = !testbed->noSleep;
        R2ColliderDesc collider = r2CuboidColliderDesc(r2Vector(0.5, 5));

        R2RigidBodyHandle rigidBodyHandle = r2InsertRigidBody(world, &rigidBody);
        r2InsertCollider(rigidBodyHandle, &collider);
    }
    {
        R2RigidBodyDesc rigidBody = r2FixedRigidBodyDesc();
        rigidBody.position.translation = r2Vector(15, 5);
        rigidBody.canSleep = !testbed->noSleep;
        R2ColliderDesc collider = r2CuboidColliderDesc(r2Vector(0.5, 5));

        R2RigidBodyHandle rigidBodyHandle = r2InsertRigidBody(world, &rigidBody);
        r2InsertCollider(rigidBodyHandle, &collider);
    }

    /* Pressurized blobs of various sizes. */
    for (int i = 0; i < 5; i++) {
        const R2Real radius = 0.6 + 0.15 * i;
        R2SoftBodyDesc blob = r2DiskSoftBodyDesc(r2Vector(-10.0 + i * 2.5, 2.0 + i), radius, 24);
        blob.material = r2UniformSoftBodyMaterial((R2SpringCoefficients){20.0, 1.0});
        blob.volumeFactor = 1.1;
        blob.selfContacts = 1;
        blob.particleMass = 0.05;
        blob.canSleep = !testbed->noSleep;
        r2InsertSoftBody(world, &blob);
    }

    /* Jelly bodies: corotational, Neo-Hookean, and per-cell volume constraints. */
    {
        R2SoftBodyDesc jelly = r2GridSoftBodyDesc(r2Vector(2, 1.2), r2Vector(1, 1), 6, 6);
        jelly.cellModel = R2_SOFT_CELL_COROTATIONAL;
        R2SoftBodyMaterial material = r2DefaultSoftBodyMaterial();
        material.youngModulus = 3.0e3;
        material.poissonRatio = 0.35;
        material.elasticDampingRatio = 0.5;
        jelly.material = material;
        jelly.particleMass = 0.2;
        jelly.canSleep = !testbed->noSleep;
        r2InsertSoftBody(world, &jelly);
    }
    {
        R2SoftBodyDesc jelly = r2GridSoftBodyDesc(r2Vector(8, 1.2), r2Vector(1, 1), 6, 6);
        jelly.cellModel = R2_SOFT_CELL_NEO_HOOKEAN;
        R2SoftBodyMaterial material = r2DefaultSoftBodyMaterial();
        material.youngModulus = 3.0e3;
        material.poissonRatio = 0.35;
        material.elasticDampingRatio = 0.5;
        jelly.material = material;
        jelly.particleMass = 0.2;
        jelly.canSleep = !testbed->noSleep;
        r2InsertSoftBody(world, &jelly);
    }
    {
        R2SoftBodyDesc jelly = r2GridSoftBodyDesc(r2Vector(5, 1.2), r2Vector(1, 1), 6, 6);
        jelly.cellModel = R2_SOFT_CELL_VOLUME;
        jelly.material = r2UniformSoftBodyMaterial((R2SpringCoefficients){20.0, 1.0});
        jelly.particleMass = 0.2;
        jelly.canSleep = !testbed->noSleep;
        r2InsertSoftBody(world, &jelly);
    }

    /* A rope hanging from a fixed anchor, holding a rigid weight. */
    const uint32_t pinnedParticle = 0;
    R2SoftBodyDesc rope = r2DefaultSoftBodyDesc();
    rope.kind = R2_SOFT_DESC_ROPE;
    rope.a = r2Vector(8, 9);
    rope.b = r2Vector(12, 9);
    rope.nx = 25;
    rope.pinned = (R2IndexView){&pinnedParticle, 1};
    rope.material.edgeSoftness = rope.material.bendSoftness = rope.material.volumeSoftness =
        rope.material.shapeMatchingSoftness = (R2SpringCoefficients){40.0, 1.0};
    rope.particleMass = 0.05;
    rope.canSleep = !testbed->noSleep;
    R2SoftBodyHandle ropeHandle = r2InsertSoftBody(world, &rope);

    R2Vector lastPos = r2SoftBody_ParticlePosition(ropeHandle, 24);
    R2RigidBodyHandle weight;
    {
        R2RigidBodyDesc rigidBody = r2DynamicRigidBodyDesc();
        rigidBody.position.translation = r2VectorAdd(lastPos, r2Vector(0, -0.4));
        rigidBody.canSleep = !testbed->noSleep;
        R2ColliderDesc collider = r2CuboidColliderDesc(r2Vector(0.3, 0.3));
        collider.density = 2.0;
        weight = r2InsertRigidBody(world, &rigidBody);
        r2InsertCollider(weight, &collider);
    }

    r2SoftBody_AttachParticle(ropeHandle, 24, weight);

    /* A stack of rigid boxes for the blobs to knock down. */
    for (int i = 0; i < 6; i++) {
        R2RigidBodyDesc rigidBody = r2DynamicRigidBodyDesc();
        rigidBody.position.translation = r2Vector(-4.0, 0.3 + 0.6 * i);
        rigidBody.canSleep = !testbed->noSleep;
        R2ColliderDesc collider = r2CuboidColliderDesc(r2Vector(0.3, 0.3));

        R2RigidBodyHandle rigidBodyHandle = r2InsertRigidBody(world, &rigidBody);
        r2InsertCollider(rigidBodyHandle, &collider);
    }

    /* Set up the viewer. */
    tbCamera2(testbed, 0.0, 4.0, 30.0);

    /* Set up rendering and run the simulation. */
    tbSetWorld(testbed, world);

    while (tbRenderFrame(testbed, &world)) {
        if (tbSimulating(testbed)) {
            r2Step(world, NULL, NULL);
        }
    }
    r2FreeWorld(world);
}
